### PR TITLE
remove unneeded transpose in matmul

### PR DESCRIPTION
**Pull request by oxinabox**
I suspect I am doing something silly here, but i can't see why we would want a transpose when we can just change the order of the indexes.

### CODE REVIEW

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Looks like `main` is your base. Could you please rebase on `dev`?

File: prelude.dx [hunk 1]
```diff
@@ -419,8 +419,7 @@ def vdot (x:n=>Float) (y:n=>Float) : Float = fsum \i. x.i * y.i
 
 -- matmul. Better symbol to use? `@`?
 (**) : (l=>m=>Float) -> (m=>n=>Float) -> (l=>n=>Float) = \x y.
-  y' = transpose y
-  for i k. fsum \j. x.i.j * y'.k.j
+  for i k. fsum \j. x.i.j * y.j.k
```
Comment (apaszke) [inline prelude.dx:422]: BTW another potential improvement is replacing `fsum \j.` with `sum for j.`. They should be equally fast given that we've added some optimizations.

Comment (oxinabox) [reply prelude.dx:473]: should we remove fusm everywhere and remove that function?
For now i will do so just from transpose

**Comment by apaszke** [comment]
Yeah, I'd be happy to get rid of `fsum` all over the place, but it's not blocking for this PR

**Comment by dougalm** [comment]
> i can't see why we would want a transpose when we can just change the order of the indexes.

It was a slightly embarrassing performance hack. Matrix multiplication can be faster with some data layouts than others (e.g. column-major instead of row-major). In principle, the compiler is free to choose layout however it wants, irrespective of the logical ordering of indices. But, until Adam implemented inlining of `for` expressions, the layout was always row-major. The transpose here  forced a physical transpose in memory to produce the desired layout.

**Comment by dan-zheng** [comment]
`dev` branch has been merged into `main`, now that plotting has been revamped. `dev` branch will be deleted now - we can continue development on `main`.

I'll change the base branch of this PR to `main` now!

**Comment by axch** [comment]
The changed in this PR have both happened, so I'll close it.

**Pull request closed without merging**